### ISSUE TITLE
Are composed mutations currently supported?

### ISSUE THREAD

**Issue by gdw2** [retro/graphql-builder#14]
Are composed mutations currently supported?

I'm currently building a composed query out of mutations.  I _think_ the resulting query will work as a mutation if I `s/query/mutation` in the resulting query string.

**Comment by retro** [comment]
I see that we're throwing an exception when you try t compose mutations, but frankly I don't remember why. I think that the reasoning was that the results could be unexpected, because mutations could be ran in non deterministic order. On the other hand, if you need this, you're probably aware of these potential issues.

Do you have an example mutation you would like to compose? I could use it to write tests

**Comment by gdw2** [comment]
Sorry to leave you hanging for so long.  I'll try to get you a PR.

Interestingly, I can compose mutations. It doesn't throw an exception.  It just uses the word `query` instead of `mutation`.  So, at present, I'm just doing a 'replace' on the query string (`(clojure.string/replace (get-in query [:graphql :query]) #"^query" "mutation")`) and everything works as expected.

**Comment by retro** [comment]
If that’s so, I think we should remove this check and allow mutation composition without checks. If you can send me an example it would be great, otherwise I’ll write some tests and push new release

**Comment by gdw2** [comment]
https://github.com/retro/graphql-builder/pull/17

**Comment by mharju** [comment]
We are currently experimenting with composing mutations. We managed to do the proper implementation by just doing a few new functions:

```clojure
(defn make-mutation [mutation-name composition-parts]
  (-> [(str "mutation " mutation-name (gqlcomp/query-variables composition-parts) " {")
       (map :children (vals composition-parts))
       "}"]
      flatten
      gqlu/nl-join))

(defn generate-mutation [visitor mutations nodes]
  (let [mutation-name "ComposedMutation"
        prefixes (gqlcomp/make-prefixes mutations)
        mutation-nodes (gqlcomp/make-query-nodes nodes mutations)
        composition-parts (gqlcomp/make-composition-parts visitor mutation-nodes prefixes)
        mutation (make-mutation mutation-name composition-parts)]
    (fn op-fn
      ([] (op-fn {}))
      ([vars]
       (let [namespaced-vars (gqlcomp/namespace-vars prefixes vars)]
         {:graphql {:operationName mutation-name
                    :query mutation
                    :variables namespaced-vars}
          :unpack (gqlcomp/make-unpack (gqlu/reverse-map prefixes))})))))

(defn composed-mutation [parsed-statement mutations]
  (let [nodes (gqlc/generate parsed-statement {:generate? false})]
    (generate-mutation gqlc/generate-visit-nodes mutations nodes)))
```

This could be added to the compose-namespace directly, add some tests and perhaps refactor a bit. We'll get back to this as we try this out and see how it goes.

**Comment by retro** [comment]
@mharju Thank you, that would be great

**Comment by retro** [comment]
@mharju please check this pull request https://github.com/retro/graphql-builder/pull/17 so it works in that case too

**Comment by mharju** [comment]
OK, I pushed #19 to implement this. If you have any comments, please let me know and I'll adjust accordingly.

We have tested this implementation in our current project in with Postgraphile in the backend and it seems to be working as expected.

**Comment by retro** [comment]
Merged in d04953cf4fb76d868ab87c2308e034e7ecc34b55
